devices/flash-zynq7000/qspi.c

```diff
@@ -0,0 +1,329 @@
+			if ((hal_timerGet() - start) >= timeout)
+				return -ETIME;
+		}
+
+		/* Receive data */
+		while (rxSz) {
+			/* RX Fifo is empty */
+			if ((*(qspi_common.base + sr) & (1 << 4)) == 0)
+				break;
+
+			tempSz = qspi_rxData(rxBuff, rxSz);
+
+			rxSz -= tempSz;
+			rxBuff += tempSz;
+		}
+
+		if ((hal_timerGet() - start) >= timeout)
+			return -ETIME;
+	}
+
+	return size - txSz;
+}
+
+
+static int qspi_setPin(u32 pin)
+{
+	ctl_mio_t ctl;
+
+	ctl.pin = pin;
+	ctl.l0 = 0x1;
+	ctl.l1 = ctl.l2 = ctl.l3 = 0;
+	ctl.pullup = 0;
+	ctl.speed = 0x1;
+	ctl.ioType = 0x1;
+	ctl.disableRcvr = 0;
+	ctl.triEnable = 0;
+
+	switch (pin) {
+		/* Chip select */
+		case mio_pin_01:
+			ctl.pullup = 1;
+			ctl.speed = 0x0;
+			break;
+
+		case mio_pin_02:
+		case mio_pin_03:
+		case mio_pin_04:
+		case mio_pin_05:
+		case mio_pin_06:
+			break;
+
+		default:
+			return -EINVAL;
+	}
+
+	return _zynq_setMIO(&ctl);
+}
+
+
+static int qspi_initCtrlClk(void)
+{
+	ctl_clock_t ctl;
+
+	/* Set IO PLL as source clock and set divider:
+	 * IO_PLL / 0x5 :  1000 MHz / 5 = 200 MHz     */
+	ctl.dev = ctrl_lqspi_clk;
+	ctl.pll.clkact0 = 0x1;
+	ctl.pll.clkact1 = 0x1;
+	ctl.pll.srcsel = 0;
+	ctl.pll.divisor0 = 5;
+
+	return _zynq_setCtlClock(&ctl);
+}
+
+
+/* Linear mode allows only for reading data.
+ * 03h command is recommended, otherwise first word = 0 (internal bug) :
+ * https://support.xilinx.com/s/article/60803?language=en_US
+ */
+#if 0
+static int qspi_linearMode(void)
+{
+	/* Disable QSPI */
+	*(flashdrv_common.base + er) &= ~0x1;
+
+	/* Disable IRQs */
+	*(flashdrv_common.base + idr) = 0x7d;
+
+	/* Disable linear mode */
+	*(flashdrv_common.base + lqspi_cr) = 0;
+
+	*(flashdrv_common.base + cr) = (1 << 14);
+	*(flashdrv_common.base + cr) = (1 << 10);
+
+	*(flashdrv_common.base + cr) &= ~(0x7 << 3);
+	*(flashdrv_common.base + cr) |= (0x3 << 3);
+
+	*(flashdrv_common.base + cr) |= 0x1;
+	*(flashdrv_common.base + cr) |= (1 << 31);
+	*(flashdrv_common.base + cr) &= ~(1 << 26);
+
+	*(flashdrv_common.base + cr) |= (0x3 << 6);
+	*(flashdrv_common.base + cr) &= ~(0x3 << 1);
+	*(flashdrv_common.base + cr) |= (0x1 << 19);
+
+	*(flashdrv_common.base + lqspi_cr) =  0x80000003;
+
+	*(flashdrv_common.base + er) = 0x1;
+
+	return EOK;
+}
+#endif
+
+
+static void qspi_IOMode(void)
+{
+	/* Configure I/O mode */
+
+	/* Disable QSPI */
+	*(qspi_common.base + er) &= ~0x1;
+	/* Disable IRQs */
+	*(qspi_common.base + idr) = 0x7d;
+
+	/* Configure controller */
+
+	/* Set baud rate, master mode, not Legacy mode */
+	*(qspi_common.base + cr) &= ~(0x1 << 3);
+	*(qspi_common.base + cr) |= (0x1 << 3) | 0x1 | (1 << 31);
+	/* Set little endian */
+	*(qspi_common.base + cr) &= ~(1 << 26);
+	/* Set FIFO width 32 bits */
+	*(qspi_common.base + cr) |= (0x3 << 6);
+	/* Set clock phase and polarity */
+	*(qspi_common.base + cr) &= ~(0x3 << 1);
+
+
+	/* Enable manual mode and manual CS */
+	*(qspi_common.base + cr) |= (0x3 << 14);
+
+	/* Disable linear mode */
+	*(qspi_common.base + lqspi_cr) = 0;
```

Comment:
And memory barier after this one. Maybe some are needed in between (I think not)?